### PR TITLE
rust/ldap: udp and frames support v1

### PR DESCRIPTION
**Pull request by glongo**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)
- [x] I have created a ticket at
      https://redmine.openinfosecfoundation.org/projects/suricata/issues
      (if applicable)

Link to ticket: https://redmine.openinfosecfoundation.org/issues/1199

Describe changes:
- Enable parser and logger for LDAP traffic over UDP
- Add PDU frames

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1984

### CODE REVIEW

File: rust/src/ldap/ldap.rs [hunk 1]
```diff
@@ -426,4 +426,24 @@ pub unsafe extern "C" fn rs_ldap_register_parser() {
...
+
+    parser.ipproto = IPPROTO_UDP;
+    let ip_proto_str = CString::new("udp").unwrap();
+    if AppLayerProtoDetectConfProtoDetectionEnabled(ip_proto_str.as_ptr(), parser.name) != 0 {
+        let alproto = AppLayerRegisterProtocolDetection(&parser, 1);
+        ALPROTO_LDAP = alproto;
+        if AppLayerParserConfParserEnabled(ip_proto_str.as_ptr(), parser.name) != 0 {
+            let _ = AppLayerRegisterParser(&parser, alproto);
+        }
+        if let Some(val) = conf_get("app-layer.protocols.ldap.max-tx") {
```
Comment (catenacyber) [inline rust/src/ldap/ldap.rs:499]: This code should only run once

File: src/output.c [hunk 2]
```diff
@@ -1094,7 +1094,7 @@ void OutputRegisterLoggers(void)
             JsonLogThreadDeinit, NULL);
     /* Ldap JSON logger. */
     OutputRegisterTxSubModule(LOGGER_JSON_TX, "eve-log", "JsonLdapLog", "eve-log.ldap",
-            OutputJsonLogInitSub, ALPROTO_LDAP, JsonGenericDirPacketLogger, JsonLogThreadInit,
+            OutputJsonLogInitSub, ALPROTO_LDAP, JsonGenericDirFlowLogger, JsonLogThreadInit,
```
Comment (catenacyber) [inline src/output.c:1097]: Nice catch, bidirectional transactions should log with flow indeed

File: rust/src/ldap/ldap.rs [hunk 3]
```diff
@@ -426,4 +426,24 @@ pub unsafe extern "C" fn rs_ldap_register_parser() {
     } else {
         SCLogDebug!("Protocol detection and parser disabled for LDAP.");
     }
+
+    parser.ipproto = IPPROTO_UDP;
```
Comment (catenacyber) [inline rust/src/ldap/ldap.rs:491]: Should not there be an update to schema.json with `ldap_udp` somewhere ?

File: rust/src/ldap/ldap.rs [hunk 4]
```diff
@@ -32,6 +33,13 @@ static mut LDAP_MAX_TX: usize = 256;
 
 static mut ALPROTO_LDAP: AppProto = ALPROTO_UNKNOWN;
 
+// app-layer-frame-documentation tag start: FrameType enum
```
Comment (catenacyber) [inline rust/src/ldap/ldap.rs:36]: We do not need this line, it is used by documentation fro showcase the sip one

File: rust/src/ldap/ldap.rs [hunk 5]
```diff
@@ -32,6 +33,13 @@ static mut LDAP_MAX_TX: usize = 256;
...
+#[derive(AppLayerFrameType)]
+pub enum LdapFrameType {
+    Pdu,
```
Comment (catenacyber) [inline rust/src/ldap/ldap.rs:39]: Could we have header and data ? Or is it not relevant for LDAP ?

File: rust/src/ldap/ldap.rs [hunk 6]
```diff
@@ -158,24 +170,43 @@ impl LdapState {
...
+        let input = stream_slice.as_slice();
         if input.is_empty() {
             return AppLayerResult::ok();
         }
 
+        let _pdu = Frame::new(
+            flow,
+            &stream_slice,
+            input,
+            input.len() as i64,
```
Comment (catenacyber) [inline rust/src/ldap/ldap.rs:183]: Rather -1 to start, no ?

File: rust/src/ldap/ldap.rs [hunk 7]
```diff
@@ -158,24 +170,43 @@ impl LdapState {
...
                     tx.complete = match request.protocol_op {
                         ProtocolOp::UnbindRequest => true,
                         _ => false,
                     };
                     tx.request = Some(request);
                     self.transactions.push_back(tx);
+
+                    let consumed = start.len() - rem.len();
+                    start = rem;
+                    if let Some(frame) = &self.request_frame {
```
Comment (catenacyber) [inline rust/src/ldap/ldap.rs:205]: This creates only one frame in the case there are 2 PDUs in a single TCP packet, does it not ?

File: rust/src/ldap/ldap.rs [hunk 8]
```diff
@@ -214,20 +255,44 @@ impl LdapState {
                             | ProtocolOp::ExtendedResponse(_) => true,
                             _ => false,
                         };
+                        let tx_id = tx.id();
                         tx.responses.push_back(response);
+                        let consumed = start.len() - rem.len();
+                        start = rem;
```
Comment (catenacyber) [inline rust/src/ldap/ldap.rs:261]: Could be done only once after these conditions

File: rust/src/ldap/ldap.rs [hunk 9]
```diff
@@ -316,14 +381,12 @@ unsafe extern "C" fn SCLdapParseRequest(
         }
     }
     let state = cast_pointer!(state, LdapState);
-    state.parse_request(stream_slice.as_slice());
-
-    AppLayerResult::ok()
```
Comment (catenacyber) [inline rust/src/ldap/ldap.rs:321]: Thanks for fixing this

File: rust/src/ldap/ldap.rs (hunk 3 above)
Comment (glongo) [reply rust/src/ldap/ldap.rs:491]: Right, I edited the file but forgot to include it in the commit.

File: rust/src/ldap/ldap.rs (hunk 5 above)
Comment (glongo) [reply rust/src/ldap/ldap.rs:39]: LDAP doesn't have a header per se, but just a message ID.

Comment (catenacyber) [reply rust/src/ldap/ldap.rs:39]: Where are the operation and pdu length encoded ?

**Review by catenacyber — CHANGES_REQUESTED** [review]
Thanks for this, CI needs to get green

File: rust/src/ldap/ldap.rs (hunk 7 above)
Comment (glongo) [reply rust/src/ldap/ldap.rs:205]: Yes it does.

File: rust/src/ldap/ldap.rs (hunk 5 above)
Comment (glongo) [reply rust/src/ldap/ldap.rs:39]: They are encoded following ASN.1 syntax

**Comment by glongo** [comment]
Replaced with https://github.com/OISF/suricata/pull/11534

File: rust/src/ldap/ldap.rs (hunk 5 above)
Comment (catenacyber) [reply rust/src/ldap/ldap.rs:39]: Like BER or DER encoding.
They have at least some tag stuff that we could define as the header...

But ok just to have `Pdu` for first version, thanks :-)

File: rust/src/ldap/ldap.rs (hunk 7 above)
Comment (catenacyber) [reply rust/src/ldap/ldap.rs:205]: It seems to me that `Frame::new` is outside the loop

**Pull request closed without merging**